connections/ethereum/connection.go

```diff
@@ -165,6 +165,34 @@ func (c *Connection) SafeEstimateGas(ctx context.Context) (*big.Int, error) {
 	}
 }
 
+func (c *Connection) SafeEstimateGasTipCap(ctx context.Context, baseFee *big.Int) (*big.Int, *big.Int, error) {
+	if c.maxGasPrice.Cmp(baseFee) < 0 {
+		return nil, nil, fmt.Errorf("maxGasPrice (%v) < BaseFee (%v)", c.maxGasPrice, baseFee)
+	}
+
+	var tip *big.Int
+	tip, err := c.conn.SuggestGasTipCap(context.TODO())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	gasFeeCap := new(big.Int).Add(
+		tip,
+		new(big.Int).Mul(baseFee, big.NewInt(2)),
+	)
+
+	if gasFeeCap.Cmp(tip) < 0 {
+		return nil, nil, fmt.Errorf("maxFeePerGas (%v) < maxPriorityFeePerGas (%v)", c.opts.GasFeeCap, c.opts.GasTipCap)
+	}
+
+	// Check we aren't exceeding our limit
+	if gasFeeCap.Cmp(c.maxGasPrice) == 1 {
```

Comment:
i do not understand this calculations... could you explain please

Reply:
With legacy txs if the max gas suggested is greater than the `maxGasPrice` in the config we just return the config's `maxGasPrice`, but with London we have to specify both the tip and total max cap. So I made it that the `gasTipCap` is just then the `max - baseFee`, with the `gasFeeCap` being the `maxGasPrice`